### PR TITLE
Allow ArrayRef as kernel parameter

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34399 Allow std::array as operator argument and return
* **#34335 Allow ArrayRef as kernel parameter**

Differential Revision: [D20296841](https://our.internmc.facebook.com/intern/diff/D20296841/)

### CODE REVIEW

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h [hunk 1]
```diff
@@ -99,6 +99,12 @@ namespace impl {
     // TODO static_assert(AllowDeprecatedTypes, "You tried to register a kernel with an unsupported input type: std::vector<T>. Please use List<T> instead.");
   };
 
+  template<class T, bool AllowDeprecatedTypes>
+  struct assert_is_valid_input_type<c10::ArrayRef<T>, AllowDeprecatedTypes>
+  : assert_is_valid_input_type<T, AllowDeprecatedTypes> {
+    static_assert(!std::is_same<T, at::Scalar>::value, "You tried to register a kernel with an unsupported input type: ArrayRef<Scalar>. Please use List<int64_t>, List<double> or Tensor instead.");
```
Comment (ezyang) [inline aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:105]: None of these suggested alternatives seem particularly good

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h [hunk 2]
```diff
@@ -187,12 +193,22 @@ namespace impl {
...
   template<class T, bool AllowDeprecatedTypes>
-  T ivalue_to_arg(IValue&& v) {
-    assert_is_valid_input_type<T, AllowDeprecatedTypes>();
-    return std::move(v).to<T>();
-  }
+  struct ivalue_to_arg<ArrayRef<T>, AllowDeprecatedTypes> final {
+    // If an argument is ArrayRef<T>, convert the IValue to a std::vector<T> and pass that
+    // to the operator. std::vector<T> is implicitly convertible to ArrayRef<T>.
+    static std::vector<T> call(IValue&& v) {
+      return ivalue_to_arg<std::vector<T>, AllowDeprecatedTypes>::call(std::move(v));
```
Comment (ezyang) [inline aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:209]: This sounds really slow. cc @zdevito (I guess this is a consequence of uniform List representation meaning that conversion to an ArrayRef is not O(1)? That sucks!)

**Review by ezyang — APPROVED** [review]
There are a bunch of places where people have written kernels that take `std::vector<int64_t>` when they should take array ref. Those should get updated too.

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:105]: What would you propose instead? We don't support lists of scalars afaik, so they would have to use int or double lists.

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h (hunk 2 above)
Comment (smessmer) [reply aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:209]: yes that's a consequence of that. In light of that, I think we should consider only use `ArrayRef` as a convenience feature for kernels that need the memory layout of an `ArrayRef`, for example for passing a pointer to an `ArrayRef` with strides to a CUDA kernel, but everything else should use `torch::List` which avoids the copy.

Comment (ezyang) [reply aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:209]: @smessmer I think that's a pretty big step to take, so I'd like to see more evidence for this recommendation. For example, the C++ API takes ArrayRef as a public API. I can't see how ArrayRef O(1) converts to torch::List.

File: aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:105]: I would tell them to file a bug and explain their use case :)

**Pull request closed without merging**